src/algebra/monoid_algebra/grading.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import algebra.monoid_algebra.to_direct_sum
+import linear_algebra.finsupp
+
+/-!
+# Internal grading of an `add_monoid_algebra`
+
+In this file, we show that an `add_monoid_algebra` has an internal direct sum structure.
+
+## Main results
+
+* `add_monoid_algebra.grade R i`: the `i`th grade of an `add_monoid_algebra R ι`
+* `add_monoid_algebra.equiv_grade`: the equivalence between an `add_monoid_algebra` and the direct
+  sum of its grades.
+* `add_monoid_algebra.grade.is_internal`: propositionally, the statement that
+  `add_monoid_algebra.grade` defines an internal graded structure.
+-/
+
+noncomputable theory
+
+namespace add_monoid_algebra
+variables {ι : Type*} {R : Type*}
+
+section
+variables (R)
+
+/-- The submodule corresponding to each grade. -/
+abbreviation grade [comm_semiring R] (i : ι) : submodule R (add_monoid_algebra R ι) :=
+(finsupp.lsingle i).range
+
+end
+
+variables {R} [decidable_eq ι] [add_monoid ι] [comm_semiring R]
+
+open_locale direct_sum
+
+instance grade.gmonoid : direct_sum.gmonoid (λ i : ι, grade R i) :=
+direct_sum.gmonoid.of_submodules (grade R) ⟨1, rfl⟩ $ λ i j, begin
+  rintros ⟨-, a, rfl⟩ ⟨-, b, rfl⟩,
+  exact ⟨_, single_mul_single.symm⟩,
+end
```

Comment:
So this PR takes the point of view that the "basic thing" when it comes to a collection of R-submodules of some R-algebra indexed by an `add_monoid` is `direct_sum.gmonoid`. However there is a more basic object, namely the predicate that says that this is an *internal* grading.

Reply:
```
variables {ι : Type*} [add_monoid ι]
  {R : Type*} [comm_ring R]
  {M : Type*} [add_comm_monoid M] [module R M]

open_locale direct_sum

def is_internal_direct_sum [decidable_eq ι] (G : ι → submodule R M) : Prop :=
function.bijective
  (direct_sum.to_add_monoid (λ i, (G i).subtype.to_add_monoid_hom) : (⨁ i, G i) → M)
```

Reply:
My claim is that when showing `has_add_submonoid_decomposition` (which is the hard bit), it's actually easier to have the `gmonoid` instance already available before trying to prove it. The reason I claim this is that you then can apply various universal properties that require an algebra structure on the codomain to construct the `components` map (eg `free_algebra.lift`), and can use corresponding ext lemmas to hopefully make the left and right inverse proofs a little easier.

My intent in this PR was to provide the most useful object (the isomorphism itself), such that we can then try factoring pieces (like the ones you link to in lean-liquid) out of the construction, and see if they make things easier or harder (in a way that the git diff hopefully helps us see!).